flytekit/remote/remote.py

```diff
@@ -2037,6 +2040,28 @@ def get_extra_headers_for_protocol(native_url):
             return {"x-ms-blob-type": "BlockBlob"}
         return {}
 
+    @staticmethod
+    def get_extra_headers_for_signed_url(signed_url):
+        """
+        A limitation of S3 Bucket Policies is that when enforcing SSE, checks are
+        done on the headers of the request. Therefore, when using pre-signed urls
+        we should fill in the headers with the SSE information from the url itself.
+
+        This is possible since these self-contained pre-signed urls contain the SSE
+        information within its query string.
+        """
+        headers = dict()
+        query_params = parse_qs(signed_url)
+        if "x-amz-server-side-encryption" in query_params:
+            sse = query_params["x-amz-server-side-encryption"]
+            if len(sse) > 0:
+                headers["x-amz-server-side-encryption"] = sse[0]
```

Comment:
I think what's going on is - yes, pre-signed URLs don't need headers.

But, in this case, we are specifying that the buckets should require SSE-KMS (consistency with other buckets in our deployment) and this check is usually done via headers.

* Option 1) [this PR] S3 bucket policy to **check the headers** - this is consistent with other buckets, but requires some finagling of the URL+headers as seen in this method signature
* Option 2) S3 bucket policy that **checks either the URL or the headers**, this would be ideal I think, and we had discussed this earlier. In reality, I found that S3 bucket policies are not this flexible – so option 1 was taken.
* Option 3) S3 bucket policy that **doesn't check URL or headers** – this seems insecure and its also very inconsistent with the other buckets in our deployment. But I believe uploads will be encrypted by default anyways so this isn't that bad.